### PR TITLE
Create serving certs for headless services on OpenShift (#818)

### LINKED ISSUE #818: Create serving certs for headless services on OpenShift
When running on OpenShift, the opentelemetry-operator should include a `service.beta.openshift.io/serving-cert-secret-name` annotation in the headless services it creates for the collector. This will cause a certificate to be generated for the service and stored in a secret that can be mounted and used in the associated receiver TLS configuration. 

Without this feature in the operator, it is necessary to either manage serving certificates manually, or add the annotation manually after deploying the collector. Until the annotation is added, the collector cannot start successfully because the secret being mounted as a volume does not yet exist.

### PR DESCRIPTION
**Pull request by rkukura**
resolves #818 

Signed-off-by: Robert Kukura <rkukura@redhat.com>

### CODE REVIEW

File: pkg/collector/reconcile/service.go [hunk 1]
```diff
@@ -158,6 +158,16 @@ func headless(ctx context.Context, params Params) *corev1.Service {
 	}
 
 	h.Name = naming.HeadlessService(params.Instance)
+
+	// copy to avoid modifying params.Instance.Annotations
+	annotations := map[string]string{
+		"service.beta.openshift.io/serving-cert-secret-name": fmt.Sprintf("%s-tls", h.Name),
```
Comment (pavolloffay) [inline pkg/collector/reconcile/service.go:164]: This annotation should be present only if the collector is deployed on the OpenShift. 

IIRC there is already some machinery in the project that recognizes the platform.

Comment (rkukura) [reply pkg/collector/reconcile/service.go:164]: I had considered this, but noticed in https://github.com/jaegertracing/jaeger-operator/blob/main/pkg/service/collector.go#L28 that jaeger-operator adds the annotation regardless of the platform. My understanding is that unrecognized annotations are silently ignored, so I don't think there is any harm. Also, making this conditional would require addition unit test code. But I can make it conditional if required.

**Comment by pavolloffay** [comment]
@rkukura is there any documentation needed for this? Not familiar users might ask how to use those certs? Why does only the headless service have them?

**Comment by rkukura** [comment]
> @rkukura is there any documentation needed for this? Not familiar users might ask how to use those certs? Why does only the headless service have them?

@pavolloffay I added some text to README.md. The headless service is what is typically used within the cluster, such as for connections from agents. The generated certificate is only trusted within the cluster.

File: README.md [hunk 2]
```diff
@@ -57,6 +57,8 @@ The `config` node holds the `YAML` that should be passed down as-is to the under
 
 At this point, the Operator does *not* validate the contents of the configuration file: if the configuration is invalid, the instance will still be created but the underlying OpenTelemetry Collector might crash.
 
+The Operator does examine the configuration file to discover configured receivers and their ports. If it finds receivers with ports, it creates a pair of kubernetes services, one headless, exposing those ports within the cluster. The headless service contains a `service.beta.openshift.io/serving-cert-secret-name` annotation that will cause OpenShift to create a secret containing a certificate and key. This secret can be mounted as a volume and the certificate and key used in those receivers' TLS configurations.
```
Comment (pavolloffay) [inline README.md:60]: The text looks good/ it explains what the annotation does. I would it makes sense here also document why someone should be doing something like this/waht is the common use case? Users are often searching for solutions for use-cases.

Comment (rkukura) [reply README.md:60]: More general discussion of configuring and securing the collector is not specific to the operator, so probably doesn't belong in this repo. We could add some mention of this operator functionality in https://opentelemetry.io/docs/collector/configuration/#setting-up-certificates after this PR is merged. Adding new documentation in this repo that covers common operator use cases (including but not limited to securing OTLP between agent and gateway collecotrs) seems beyond the scope of this PR.

Comment (pavolloffay) [reply README.md:60]: > More general discussion of configuring and securing the collector is not specific to the operator, so probably doesn't belong in this repo. 

Our users should understand why there is a headless service, when to use it and what is the use case for using those certificates.

> We could add some mention of this operator functionality in https://opentelemetry.io/docs/collector/configuration/#setting-up-certificates

I don't think that operator documentation belongs there. You could try to work with the docs/website sig to add an operator page there. So far there are no operator docs in opentelemetry.io.

Comment (pavolloffay) [reply README.md:60]: Maybe I was not clear, I wanted a better docs for the certificates - what is the use-case for them etc.

Comment (pavolloffay) [reply README.md:60]: You can create an additional PR to add them

**Review by pavolloffay — APPROVED** [review]

**Pull request merged**